File: CSharp/cs_EasyMKT-master/EasyMKT/Fields.cs
/* Copyright 2017. Bloomberg Finance L.P.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:  The above
 * copyright notice and this permission notice shall be included in all copies
 * or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

using Bloomberglp.Blpapi;
using System;
using System.Collections;
using System.Collections.Generic;
using static com.bloomberg.mktdata.samples.Log;

namespace com.bloomberg.mktdata.samples {

    public class Fields : IEnumerable<Field> {

        private List<Field> fields = new List<Field>();

        internal Security security;
        List<FieldChange> fieldChanges;

        internal Fields(Security security) {
            this.security = security;
            LoadFields();
        }

        private void LoadFields() {

            foreach (SubscriptionField sf in security.GetSecurities().easyMKT.fields) {
                Field newField = new Field(this, sf.GetName(), "");
                this.fields.Add(newField);
            }
        }

        internal void PopulateFields(Message message) {

            Log.LogMessage(LogLevels.DETAILED, "Populate fields");
            Log.LogMessage(LogLevels.DETAILED, "Source: " + message.ToString());

            CurrentToOldValues();

            int fieldCount = message.NumElements;

            Element e = message.AsElement;

            for (int i = 0; i < fieldCount; i++) {

                Element f = e.GetElement(i);

                String fieldName = f.Name.ToString();

                Field fd = field(fieldName);

                if (fd == null) fd = new Field(this);

                fd.SetName(fieldName);

                if (!f.IsNull) fd.SetCurrentValue(f.GetValueAsString());
                else fd.SetCurrentValue("");

                Log.LogMessage(LogLevels.DETAILED, "Setting field: " + fd.Name() + "\tvalue: " + fd.Value());
            }

            fieldChanges = new List<FieldChange>();

            foreach (Field f in fields) {

                FieldChange fc = f.GetFieldChanged();

                if (fc != null) {
                    fieldChanges.Add(fc);
                    f.sendNotifications(new List<FieldChange>{ fc });
                }
            }
        }

        internal List<FieldChange> GetFieldChanges() {
            return this.fieldChanges;
        }

        internal void CurrentToOldValues() {

            foreach (Field f in fields) {
                f.CurrentToOld();
            }
        }

        public Field field(String name) {
            foreach (Field f in fields) {
                if (f.Name().Equals(name)) return f;
            }
            return null;
        }

        public IEnumerator<Field> GetEnumerator()
        {
            return fields.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return fields.GetEnumerator();
        }
    }
}

File: Python/easyioi-0.1.0/easyioi/fields.py
# fields.py

from .field import Field

class Fields:

    def __init__(self, owner):
        self.owner = owner
        self.fields = []
        self.field_changes = []

        self.load_fields()

    def Load_fields(self):

        for sf in self.owner.easyioi.subscription_fields:
            f = Field(self,sf,"")
            self.fields.append(f)
    
    def populate_fields(self,msg):

        self.current_to_old_values()

        field_count = msg.numElements()

        for i in range(0, field_count):

            f = msg.getElement(i)
            field_name = stsr(f.name())

            fd = self.field(field_name)
            if fd == None:
                fd = Field(self,field_name)
            
            if f.isNull():
                fd.set_value("")
            else:
                fd.set_value(f.getValueAsString())
        
        self.field_changes = []

        for f in self.fields:

            fc = f.get_field_changed()
            if fc is not None:
                self.field_changes.append(fc)
                f.send_notifications([fc])
    
    def current_to_old_values(self):
        for f in self.fields:
            f.current_to_old()
    
    def field(self, name):
        for f in self.fields:
            if f.name() == name:
                return f
        return None
    
    def get_field_changes(self):
        return self.field_changes


__copyright__ = """
Copyright 2017. Bloomberg Finance L.P.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to
deal in the Software without restriction, including without limitation the
rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
sell copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:  The above
copyright notice and this permission notice shall be included in all copies
or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
IN THE SOFTWARE.
"""